File: scripts/fetch_github.py
#!/usr/bin/env python3
"""Fetches public GitHub repos for tibuntu and generates docs/projects.md."""

import json
import os
import urllib.request

GITHUB_USER = "tibuntu"
OUTPUT_FILE = "docs/projects.md"


def fetch(url):
    req = urllib.request.Request(url)
    req.add_header("Accept", "application/vnd.github+json")
    req.add_header("User-Agent", "zensical-site-builder")
    token = os.environ.get("GITHUB_TOKEN")
    if token:
        req.add_header("Authorization", f"Bearer {token}")
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())


def build_markdown(repos):
    lines = [
        "# Projects",
        "",
        "Public repositories on [GitHub](https://github.com/tibuntu), "
        "automatically updated on every deploy.",
        "",
    ]

    for repo in repos:
        if repo.get("fork"):
            continue
        name = repo["name"]
        url = repo["html_url"]
        description = repo.get("description") or ""
        language = repo.get("language") or ""
        stars = repo.get("stargazers_count", 0)

        lines.append(f"## [{name}]({url})")
        if description:
            lines.append("")
            lines.append(description)
        meta = []
        if language:
            meta.append(f"**Language:** {language}")
        if stars:
            meta.append(f"**Stars:** {stars}")
        if meta:
            lines.append("")
            lines.append(" &nbsp;·&nbsp; ".join(meta))
        lines.append("")

    return "\n".join(lines)


def main():
    url = (
        f"https://api.github.com/users/{GITHUB_USER}/repos"
        "?sort=updated&per_page=100&type=public"
    )
    repos = fetch(url)
    repos = [r for r in repos if not r.get("fork")]
    repos.sort(key=lambda r: r.get("stargazers_count", 0), reverse=True)

    md = build_markdown(repos)

    os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)
    with open(OUTPUT_FILE, "w") as f:
        f.write(md)

    print(f"Written {len(repos)} repos to {OUTPUT_FILE}")


if __name__ == "__main__":
    main()

File: scripts/fetch-github.mjs
#!/usr/bin/env node
// Fetches public GitHub repos for tibuntu and writes src/data/projects.json.
// Port of the previous scripts/fetch_github.py.

import { writeFile, mkdir } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const GITHUB_USER = 'tibuntu';
const OUTPUT = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '..',
  'src',
  'data',
  'projects.json'
);

async function fetchRepos() {
  const url =
    `https://api.github.com/users/${GITHUB_USER}/repos` +
    `?sort=updated&per_page=100&type=public`;

  const headers = {
    Accept: 'application/vnd.github+json',
    'User-Agent': 'tibuntu-dev-site-builder',
  };
  if (process.env.GITHUB_TOKEN) {
    headers.Authorization = `Bearer ${process.env.GITHUB_TOKEN}`;
  }

  const res = await fetch(url, { headers });
  if (!res.ok) {
    throw new Error(`GitHub API ${res.status}: ${await res.text()}`);
  }
  return res.json();
}

function shape(repos) {
  return repos
    .filter((r) => !r.fork && !r.archived && r.name !== 'github-pages')
    .sort((a, b) => {
      const sa = a.stargazers_count ?? 0;
      const sb = b.stargazers_count ?? 0;
      if (sb !== sa) return sb - sa;
      return new Date(b.updated_at) - new Date(a.updated_at);
    })
    .map((r) => ({
      name: r.name,
      url: r.html_url,
      description: r.description ?? '',
      language: r.language ?? '',
      stars: r.stargazers_count ?? 0,
    }));
}

async function main() {
  let projects = [];
  try {
    const repos = await fetchRepos();
    projects = shape(repos);
  } catch (err) {
    console.warn(`[fetch-github] could not fetch repos: ${err.message}`);
    console.warn('[fetch-github] writing empty projects list as fallback.');
  }

  await mkdir(dirname(OUTPUT), { recursive: true });
  await writeFile(OUTPUT, JSON.stringify(projects, null, 2) + '\n');
  console.log(`[fetch-github] wrote ${projects.length} repos to ${OUTPUT}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
